### PR TITLE
TCD/JDJ/RULE19-15

### PR DESCRIPTION
**Pull request by JacksonJ-KC**
Getting schema error: "Schema validation in rule-19-15-(a,b,c,d) for the Baseline RMR: schema invalid: Additional properties are not allowed ('supply_design_heating_setpoint_temperature' was unexpected)"

So far have not been able to figure out what I am doing incorrectly.

### CODE REVIEW

**Comment by yunjoonjung-PNNL** [comment]
You can copy the ```supply_design_heating_setpoint_temperature``` key from the schema [HERE ](https://github.com/open229/ruleset-model-description-schema/blob/main/docs229/ASHRAE229.schema.json)and paste the key into the schema in your branch. 

I was going to show you the exact location, but for some reason ```copy to highlight``` doesn't work on my laptop. If it's not clear, please let me know.

**Comment by JacksonJ-KC** [comment]
> You can copy the `supply_design_heating_setpoint_temperature` key from the schema [HERE ](https://github.com/open229/ruleset-model-description-schema/blob/main/docs229/ASHRAE229.schema.json)and paste the key into the schema in your branch.
> 
> I was going to show you the exact location, but for some reason `copy to highlight` doesn't work on my laptop. If it's not clear, please let me know.

I see! So these schema updates have not been merged into the develop branch yet?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+                                                "spaces": [
+                                                    {
+                                                        "id": "Space 1",
+                                                        "lighting_space_type": "STAIRWELL"
+                                                    }
+                                                ]
+                                            }
+                                        ],
+                                        "heating_ventilating_air_conditioning_systems": [
+                                            {
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:111]: This system isn't identified as any system. To be identified as system 10, it seems like ```cooling_system``` and ```cooling_system_type: "NONE"``` are missing. Please double-check this.

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+                                                    "heating_system_type": "FURNACE",
+                                                    "energy_source_type": "NATURAL_GAS"
+                                                },
+                                                "fan_system": {
+                                                    "id": "CAV Fan System 1",
+                                                    "fan_control": "CONSTANT",
+                                                    "supply_fans": [
+                                                        {
+                                                            "id": "Supply Fan 1"
+                                                        }
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:338]: Please add ```minimum_outdoor_airflow``` key here. The key is only used in baseline, but we need it in the proposed to get through the ```required_fields``` attribute.

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+                                        ],
+                                        "heating_ventilating_air_conditioning_systems": [
+                                            {
+                                                "id": "System 9",
+                                                "heating_system": {
+                                                    "id": "Furnace Coil 1",
+                                                    "heating_system_type": "FURNACE",
+                                                    "energy_source_type": "NATURAL_GAS"
+                                                },
+                                                "fan_system": {
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:332]: Could you add the ```specification_method``` key? The key is needed in the ```get_fan_system_object_supply_return_exhaust_relief_total_power_flow``` function.

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 4]
```diff
@@ -0,0 +1,554 @@
...
+                                                    "id": "CAV Fan System 1",
+                                                    "fan_control": "CONSTANT",
+                                                    "supply_fans": [
+                                                        {
+                                                            "id": "Supply Fan 1"
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:474]: could you add the ```minimum_outdoor_airflow``` key here?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+                                                    "heating_system_type": "FURNACE",
+                                                    "energy_source_type": "NATURAL_GAS"
+                                                },
+                                                "fan_system": {
+                                                    "id": "CAV Fan System 1",
+                                                    "fan_control": "CONSTANT",
+                                                    "supply_fans": [
+                                                        {
+                                                            "id": "Supply Fan 1",
+                                                            "design_airflow": 117.98686079999996,
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:537]: Could you add ```specification_method``` and ```design_electric_power``` keys here?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 1 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:111]: I'm not sure what I should do here. It seems like all Systems 9 & 10 in the TCD Spreadsheet for section 19 leave the `cooling_system` and `cooling_system_type` blank the same as here. Is this issue for all systems 9 & 10?

Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:111]: @jugonzal07 Any idea why the template for system 10 isn't being identified here?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 5 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:549]: Added!

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 4 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:483]: Added!

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+                                                    "energy_source_type": "NATURAL_GAS"
+                                                },
+                                                "fan_system": {
+                                                    "id": "CAV Fan System 1",
+                                                    "fan_control": "CONSTANT",
+                                                    "supply_fans": [
+                                                        {
+                                                            "id": "Supply Fan 1",
+                                                            "design_airflow": 141.58423295999995,
+                                                            "is_airflow_autosized": true
```
[the remark this answers is not in the document]
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:410]: Added!

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 3 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:338]: Added!

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 1 above)
Comment (yunjoonjung-PNNL) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:112]: We discussed this thru emails. Please let me know if you have further questions. We can have a brief call and I'll show you how we can get this issue fixed.

**Comment by yunjoonjung-PNNL** [comment]
Could you check system 10? This is because the system is not found in the ```get_baseline_system_types``` function. I believe the system in the section 19 Excel master file is outdated.

**Comment by JacksonJ-KC** [comment]
> Could you check system 10? This is because the system is not found in the `get_baseline_system_types` function. I believe the system in the section 19 Excel master file is outdated.

This issue should be fixed now

**Comment by yunjoonjung-PNNL** [comment]
> > Could you check system 10? This is because the system is not found in the `get_baseline_system_types` function. I believe the system in the section 19 Excel master file is outdated.
> 
> This issue should be fixed now

Thanks for addressing the comment quickly. However, as shown in the [RDS](https://github.com/pnnl/ruleset-checking-tool/blob/develop/docs/section19/Rule19-15.md#:~:text=hvac.fan_system)%5B%22supply_fans_airflow%22%5D%20%3E%3D-,hvac.fan_system.minimum_outdoor_airflow,-and%20are_all_hvac_sys_fan_objs_autosized(B_RMI),  we must have the ```fan_system``` and ```minimum_outdoor_airflow``` keys.

Therefore, using system 9 or system 10 that has the ```fan_system``` key would be appropriate for this rule.

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 7]
```diff
@@ -0,0 +1,618 @@
...
+                                                            "id": "Supply Fan 1",
+                                                            "specification_method": "SIMPLE",
+                                                            "design_electric_power": 75,
+                                                            "design_airflow": 117.98686079999996,
+                                                            "is_airflow_autosized": true
+                                                        }
+                                                    ],
+                                                    "return_fans": [
+                                                        {
+                                                            "id": "Return Fan 1"
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:144]: Could you add ```"specification_method": "SIMPLE",``` to all the return fans (e.g., TCD (b), (c), etc. cases)? This is the required key when the ```get_fan_system_object_supply_return_exhaust_relief_total_power_flow``` or ```get_fan_object_electric_power``` functions are used. 

Further,  please add the ```design_electric_power``` key ONLY IF it's is required for the return fans.

Comment (yunjoonjung-PNNL) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json]: @JacksonJ-KC 

I just found my comment was confusing. Could you add the ```design_electric_power``` key to all the fans (e.g., supply/return, etc.)?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+                                                        "id": "Air Terminal",
+                                                        "is_supply_ducted": true,
+                                                        "type": "CONSTANT_AIR_VOLUME",
+                                                        "served_by_heating_ventilating_air_conditioning_system": "System 10",
+                                                        "heating_source": "ELECTRIC",
+                                                        "fan": {
+                                                            "id": "Terminal Fan 1"
+                                                        },
+                                                        "primary_airflow": 471.94744319999984,
+                                                        "supply_design_heating_setpoint_temperature": 105
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:110]: Could you check if the unit was set in the Excel file? This seems very high in the SI unit.

Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:110]: Corrected to Fahrenheit

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json (hunk 7 above)
Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json]: Added

**Comment by yunjoonjung-PNNL** [comment]
Could you also update this TCD to meet schema 0.0.29?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 9]
```diff
@@ -0,0 +1,646 @@
+{
+    "rule-19-15-a": {
+        "Section": 19,
+        "Rule": 15,
+        "Test": "a",
+        "test_description": "Baseline system type 10, minimum ventilation does not exceed auto-sized airflow rates. Suppy airflow rates are correctly based on the temperature difference between 105F and the design space-heating temperature set point.",
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:6]: Could you correct this```Suppy``` typoe (also in other cases)?

**Comment by JacksonJ-KC** [comment]
> Could you also update this TCD to meet schema 0.0.29?

[df12e8a](https://github.com/pnnl/ruleset-checking-tool/pull/1100/commits/df12e8aebe7c67a10dce9e1376525121fe06e2aa) is using 0.0.29. What keys are you running into issues with?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 10]
```diff
@@ -0,0 +1,646 @@
...
+                                                        }
+                                                    ],
+                                                    "return_fans": [
+                                                        {
+                                                            "id": "Return Fan 1",
+                                                            "specification_method": "SIMPLE",
+                                                            "design_electric_power": 25
+                                                        }
+                                                    ],
+                                                    "minimum_outdoor_airflow": 0.0
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:321]: To fail, this ```minimum_outdoor_airflow``` value must be greater than the ```design_airflow``` under the ```supply_fans```. Could you check this again?

**Comment by yunjoonjung-PNNL** [comment]
Could you also add undetermined cases?

File: rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json [hunk 11]
```diff
@@ -0,0 +1,646 @@
...
+                                                "thermostat_cooling_setpoint_schedule": "Required Cooling Schedule 1",
+                                                "thermostat_heating_setpoint_schedule": "Required Heating Schedule 1",
+                                                "terminals": [
+                                                    {
+                                                        "id": "Air Terminal",
+                                                        "is_supply_ducted": true,
+                                                        "type": "CONSTANT_AIR_VOLUME",
+                                                        "served_by_heating_ventilating_air_conditioning_system": "System 9",
+                                                        "primary_airflow": 471.94744319999984,
+                                                        "supply_design_heating_setpoint_temperature": 40.5555555555556
```
Comment (yunjoonjung-PNNL) [inline rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:438]: If the ```supply_design_heating_setpoint_temperature``` temp is 105 degF, the ```design_airflow``` in supply fans has to be greater than the ```minimum_outdoor_airflow``` value. However, this case was already created in (a), you may wat to set the ```supply_design_heating_setpoint_temperature``` not equal to 105 degF, and ```design_airflow``` =```minimum_outdoor_airflow```

Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json:438]: Adjusted the `supply_heating_setpoint_temperature` so the code can reach testing for Case 2.

Comment (jugonzal07) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json]: Hi @JacksonJ-KC ,

I tested your new test JSON on 19-15 and had to make a few updates to our python code, but with this fix everything appears to pass/fail/etc. as expected.

Comment (JacksonJ-KC) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json]: Excellent news!

Comment (yunjoonjung-PNNL) [reply rct229/ruletest_engine/ruletest_jsons/ashrae9012019/section19/rule_19_15.json]: Thank you both! I also verified that all the cases passed as expected.

**Pull request merged**